pydantic_ai_slim/pydantic_ai/format_prompt.py

```diff
@@ -79,31 +106,105 @@ def to_xml(self, value: Any, tag: str | None) -> ElementTree.Element:
+        mapping: Mapping[Any, Any],
+        path: str = '',
+    ) -> None:
         for key, value in mapping.items():
             if isinstance(key, int):
                 key = str(key)
             elif not isinstance(key, str):
                 raise TypeError(f'Unsupported key type for XML formatting: {type(key)}, only str and int are allowed')
-            element.append(self.to_xml(value, key))
+            element.append(self._to_xml(value, key, f'{path}.{key}' if path else key))
+
+    @classmethod
+    def _parse_data_structures(
+        cls,
+        value: Any,
+        element_names: dict[str, str] | None = None,
+        fields_map: dict[str, tuple[str, FieldInfo | ComputedFieldInfo]] | None = None,
+        path: str = '',
+    ):
+        """Parse data structures as dataclasses or Pydantic models to extract element names and attributes."""
+        if value is None or isinstance(value, (str, int, float, date, bytearray, bytes, bool)):
+            return
+        elif isinstance(value, Mapping):
+            for k, v in value.items():  # pyright: ignore[reportUnknownVariableType]
+                cls._parse_data_structures(v, element_names, fields_map, f'{path}.{k}' if path else f'{k}')
+        elif is_dataclass(value) and not isinstance(value, type):
+            if element_names is not None:
+                element_names[path] = value.__class__.__name__
+            for k, v in asdict(value).items():
+                cls._parse_data_structures(v, element_names, fields_map, f'{path}.{k}' if path else f'{k}')
+        elif isinstance(value, BaseModel):
+            if element_names is not None:
+                element_names[path] = value.__class__.__name__
+            for model_fields in (value.__class__.model_fields, value.__class__.model_computed_fields):
+                for field, info in model_fields.items():
+                    new_path = f'{path}.{field}' if path else field
+                    field_repr = f'{value.__class__.__name__}.{field}'
+                    if (fields_map is not None) and (isinstance(info, ComputedFieldInfo) or not info.exclude):
+                        fields_map[new_path] = (field_repr, info)
+                    cls._parse_data_structures(getattr(value, field), element_names, fields_map, new_path)
+        elif isinstance(value, Iterable):
+            for n, item in enumerate(value):  # pyright: ignore[reportUnknownVariableType,reportUnknownArgumentType]
+                new_path = f'{path}.[{n}]' if path else f'[{n}]'
+                cls._parse_data_structures(item, element_names, fields_map, new_path)
+
+    @classmethod
+    def _extract_attributes(cls, info: FieldInfo | ComputedFieldInfo) -> dict[str, str]:
+        attributes: dict[str, str] = {}
+        for attr in cls._FIELD_ATTRIBUTES:
+            attr_value = getattr(info, attr, None)
+            if attr_value is not None:
+                attributes[attr] = str(attr_value)
+        return attributes
```

Comment:
This can be a oneliner, so we may not need a method:

```suggestion
        return {
            attr: str(value)
            for attr in cls._FIELD_ATTRIBUTES
            if (value := getattr(info, attr, None)) is not None
        }
```